sdk/personalizer/azure-ai-personalizer/src/main/java/com/azure/ai/personalizer/implementation/util/Utility.java

```diff
@@ -0,0 +1,150 @@
+                return operationLocation.substring(operationLocation.lastIndexOf('/') + 1,
+                    operationLocation.indexOf('?'));
+            }
+        }
+        throw LOGGER.logExceptionAsError(
+            new RuntimeException("Failed to parse operation header for result Id from: " + operationLocation));
+    }
+
+
+    public static HttpPipeline buildHttpPipeline(ClientOptions clientOptions, HttpLogOptions logOptions,
+                                                 Configuration configuration, RetryPolicy retryPolicy,
+                                                 RetryOptions retryOptions, AzureKeyCredential azureKeyCredential,
+                                                 TokenCredential tokenCredential, PersonalizerAudience audience,
+                                                 List<HttpPipelinePolicy> perCallPolicies,
+                                                 List<HttpPipelinePolicy> perRetryPolicies, HttpClient httpClient) {
+
+        Configuration buildConfiguration = (configuration == null)
+            ? Configuration.getGlobalConfiguration()
+            : configuration;
+
+        ClientOptions buildClientOptions = (clientOptions == null) ? Constants.DEFAULT_CLIENT_OPTIONS : clientOptions;
+        HttpLogOptions buildLogOptions = (logOptions == null) ? Constants.DEFAULT_LOG_OPTIONS : logOptions;
+
+        String applicationId = CoreUtils.getApplicationId(buildClientOptions, buildLogOptions);
+
+        // Closest to API goes first, closest to wire goes last.
+        final List<HttpPipelinePolicy> httpPipelinePolicies = new ArrayList<>();
+        httpPipelinePolicies.add(new AddHeadersPolicy(Constants.DEFAULT_HTTP_HEADERS));
+        httpPipelinePolicies.add(new AddHeadersFromContextPolicy());
+        httpPipelinePolicies.add(new UserAgentPolicy(applicationId, CLIENT_NAME, CLIENT_VERSION, buildConfiguration));
+        httpPipelinePolicies.add(new RequestIdPolicy());
+
+        httpPipelinePolicies.addAll(perCallPolicies);
+        HttpPolicyProviders.addBeforeRetryPolicies(httpPipelinePolicies);
+        httpPipelinePolicies.add(ClientBuilderUtil.validateAndGetRetryPolicy(retryPolicy, retryOptions));
+
+        httpPipelinePolicies.add(new AddDatePolicy());
+
+        // Authentications
+        if (tokenCredential != null) {
+            if (audience == null) {
+                audience = PersonalizerAudience.AZURE_RESOURCE_MANAGER_PUBLIC_CLOUD;
+            }
+            httpPipelinePolicies.add(new BearerTokenAuthenticationPolicy(tokenCredential,
+                audience + DEFAULT_SCOPE));
+        } else if (azureKeyCredential != null) {
+            httpPipelinePolicies.add(new AzureKeyCredentialPolicy(Constants.OCP_APIM_SUBSCRIPTION_KEY,
+                azureKeyCredential));
+        } else {
+            // Throw exception that azureKeyCredential and tokenCredential cannot be null
+            throw LOGGER.logExceptionAsError(
```

Comment:
A bit lost one this - just to double check if we want to throw when the method says logExceptionAsError

Reply:
That method logs and return the exception back again which is rethrown.